proposals/NNNN-dxil-vectors.md

```diff
@@ -0,0 +1,190 @@
+These can be stored in allocas, static globals, groupshared variables, and SSA values.
+They can be loaded from or stored to raw buffers and used as arguments to a selection
+ of element-wise intrinsic functions as well as the standard math operators.
+They cannot be used in shader signatures, constant buffers, typed buffer, or texture types.
+
+## Detailed design
+
+### Vectors in memory representations
+
+In their alloca and variable representations, vectors in DXIL will always be represented as vectors.
+Previously individual vectors would get scalarized into scalar arrays and arrays of vectors would be flattened
+ into a one-dimensional scalar array with indexing to reflect the original intents.
+Individual vectors will now be represented as a single native vector and arrays of vectors will remain
+ as arrays of native vectors, though multi-dimensional arrays will still be flattened to one dimension.
+
+Single-element vectors are generally not valid in DXIL.
+At the language level, they may be supported for corresponding intrinsic overloads,
+ but such vectors should be represented as scalars in the final DXIL output.
+Since they only contain a single scalar, single-element vectors are
+ informationally equivalent to actual scalars.
+Rather than include conversions to and from scalars and single-element vectors,
+ it is cleaner and functionally equivalent to represent these as scalars in DXIL.
+The exception is in exported library functions, which need to maintain vector representations
+ to correctly match overloads when linking.
+
+### Changes to DXIL Intrinsics
+
+A new form of rawBufferLoad allows loading of full vectors instead of four scalars.
+The status integer for tiled resource access is loaded just as before.
+The returned vector value and the status indicator are grouped into a new `ResRet` helper structure type
+ that the load intrinsic returns.
+
+```asm
+  ; overloads: SM6.9: f16|f32|i16|i32
+  ; returns: status, vector
+  declare %dx.types.ResRet.v[NUM][TY] @dx.op.rawBufferVectorLoad.v[NUM][TY](
+      i32,                  ; opcode
+      %dx.types.Handle,     ; resource handle
+      i32,                  ; coordinate c0 (byteOffset)
+      i32,                  ; coordinate c1 (elementOffset)
+      i32)                  ; alignment
+```
+
+
+The return struct contains a single vector and a single integer representing mapped tile status.
+
+```asm
+  %dx.types.ResRet.v[NUM][TY] = type { vector<TYPE, NUM>, i32 }
+```
+
+Here and hereafter, `NUM` is the number of elements in the loaded vector, `TYPE` is the element type name,
+ and `TY` is the corresponding abbreviated type name (e.g. `i64`, `f32`).
+
+#### Vector access
+
+Dynamic access to vectors were previously converted to array accesses.
+Native vectors can be accessed using `extractelement`, `insertelement`, or `getelementptr` operations.
+Previously usage of `extractelement` and `insertelement` in DXIL didn't allow dynamic index parameters.
```

Comment:
Just confirming that this spec does allow dynamic index parameters in EIE adn IEI?